onnxruntime/core/providers/cuda/controlflow/if.cc

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cuda/controlflow/if.h"
+
+#include "core/providers/cuda/cuda_common.h"
+
+//#include "core/framework/framework_common.h"
+//#include "core/framework/op_kernel_context_internal.h"
+//#include "core/framework/session_state.h"
+//#include "core/framework/tensorprotoutils.h"
+//#include "core/framework/utils.h"
+//#include "core/framework/session_options.h"
+
+using namespace ONNX_NAMESPACE;
+using namespace onnxruntime::common;
+
+namespace onnxruntime {
+namespace cuda {
+
+ONNX_OPERATOR_VERSIONED_KERNEL_EX(If,
+                                  kOnnxDomain,
+                                  1, 10,
+                                  kCudaExecutionProvider,
+                                  KernelDefBuilder()
+                                      .InputMemoryType<OrtMemTypeCPUInput>(0)  // 'cond' needs to be on CPU
+                                      .TypeConstraint("B", DataTypeImpl::GetTensorType<bool>())
+                                      .TypeConstraint("V", DataTypeImpl::AllTensorTypes()),
+                                  If);
+
+// output shape rules requiring the output shapes of the 'THEN' and 'ELSE'
+// branches to be the same were relaxed in opset-11
+ONNX_OPERATOR_KERNEL_EX(If,
+                        kOnnxDomain,
+                        11,
+                        kCudaExecutionProvider,
+                        KernelDefBuilder()
+                            .InputMemoryType<OrtMemTypeCPUInput>(0)  // 'cond' needs to be on CPU
+                            .TypeConstraint("B", DataTypeImpl::GetTensorType<bool>())
+                            .TypeConstraint("V", DataTypeImpl::AllTensorTypes()),
+                        If);
+
+Status If::Compute(OpKernelContext* ctx) const {
+  // call the base CPU version. 
+  // we have this CUDA implementation so the inputs/outputs stay on GPU where possible.
+  // the logic to run the subgraph must be on CPU either way.
+  // technically we don't need this override of Compute, but it will be optimized out and it's easier to debug 
+  // that this implementation is being called with it.
+  auto status = onnxruntime::If::Compute(ctx);
```

Comment:
What if 'cond' is on GPU ? Previously, I assume that it would have been copied over to CPU since 'If' would be assigned to CPU. But now, since 'If' will get assigned to GPU and the CPU implementation will be called, how does it handle the 'cond' available on GPU condition ?

Reply:
Ah - I just scrolled up and saw the kernel def and saw that the 'cond' had to be on CPU for an 'If' to be assigned to GPU.

Continuing on along the same lines, what happens in this scenario - 

When 'cond' is on GPU AND subgraph inputs are on GPU, it looks to me like the 'If' will get assigned to CPU and the model would incur the copy costs to CPU for all the inputs ? If so, could we just take the 'cond' input to CPU and let all the others remain on GPU and continue with the execution flow proposed by this change ? It seems like (atleast in theory) only a few models can leverage the benefit of this change - they must have 'If' node(s) with 'cond' input on CPU and some (or all) subgraph inputs on GPU and they must have GPU kernels for the corresponding inputs on GPU.

Reply:
At least in my testing 'If' will be assigned to GPU (as the CUDA EP has higher priority in our setup) and the 'cond' input will have a copy inserted to make it available on CPU due to the InputMemoryType setting. The other inputs will remain on GPU. 

There are a number of other operators that use InputMemoryType<OrtMemTypeCPUInput> like TopK, Slice, Reshape.

What makes you think If would get assigned to CPU based on the location of one input?

Reply:
My understanding of `.InputMemoryType<OrtMemTypeCPUInput>(0)` was incorrect. I thought that adding this line meant limiting assignment of If to GPU was contingent on 'cond' being available on CPU. If it means `copy cond to cpu if needed`, then I understand the flow now.